File: JasSwift/Sources/Tools/TextPathTool.swift
import AppKit
import Foundation

// MARK: - Text-on-path tool

class TextPathTool: CanvasTool {
    var dragStart: (Double, Double)?
    var dragEnd: (Double, Double)?
    var controlPt: (Double, Double)?

    func onPress(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        dragStart = (x, y)
        dragEnd = (x, y)
        controlPt = nil
    }

    func onMove(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, dragging: Bool) {
        guard let (sx, sy) = dragStart else { return }
        dragEnd = (x, y)
        let mx = (sx + x) / 2, my = (sy + y) / 2
        let dx = x - sx, dy = y - sy
        let dist = hypot(dx, dy)
        if dist > 4 {
            let nx = -dy / dist, ny = dx / dist
            controlPt = (mx + nx * dist * 0.3, my + ny * dist * 0.3)
        }
        ctx.requestUpdate()
    }

    func onRelease(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        guard let (sx, sy) = dragStart else { return }
        dragStart = nil
        dragEnd = nil
        let w = abs(x - sx), h = abs(y - sy)
        if w > 4 || h > 4 {
            let d: [PathCommand]
            if let (cx, cy) = controlPt {
                d = [.moveTo(sx, sy), .curveTo(x1: cx, y1: cy, x2: cx, y2: cy, x: x, y: y)]
            } else {
                d = [.moveTo(sx, sy), .lineTo(x, y)]
            }
            let elem = Element.textPath(JasTextPath(
                d: d, content: "Lorem Ipsum",
                fill: JasFill(color: JasColor(r: 0, g: 0, b: 0))
            ))
            ctx.controller.addElement(elem)
        }
        controlPt = nil
    }

    func drawOverlay(_ ctx: ToolContext, _ cgCtx: CGContext) {
        guard let (sx, sy) = dragStart, let (ex, ey) = dragEnd else { return }
        cgCtx.setStrokeColor(CGColor(gray: 0.4, alpha: 1.0))
        cgCtx.setLineWidth(1.0)
        cgCtx.setLineDash(phase: 0, lengths: [4, 4])
        cgCtx.move(to: CGPoint(x: sx, y: sy))
        if let (cx, cy) = controlPt {
            cgCtx.addCurve(to: CGPoint(x: ex, y: ey),
                           control1: CGPoint(x: cx, y: cy),
                           control2: CGPoint(x: cx, y: cy))
        } else {
            cgCtx.addLine(to: CGPoint(x: ex, y: ey))
        }
        cgCtx.strokePath()
        cgCtx.setLineDash(phase: 0, lengths: [])
    }
}

File: jas_ocaml/lib/text_path_tool.ml
(** Text-on-path tool: drag to create a curve with text along it. *)

class text_path_tool = object (_self)
  val mutable drag_start : (float * float) option = None
  val mutable drag_end : (float * float) option = None
  val mutable control_pt : (float * float) option = None

  method on_press (_ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(alt : bool) =
    ignore (shift, alt);
    drag_start <- Some (x, y);
    drag_end <- Some (x, y);
    control_pt <- None

  method on_move (ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(dragging : bool) =
    ignore (shift, dragging);
    match drag_start with
    | Some (sx, sy) ->
      drag_end <- Some (x, y);
      let mx = (sx +. x) /. 2.0 and my = (sy +. y) /. 2.0 in
      let dx = x -. sx and dy = y -. sy in
      let dist = sqrt (dx *. dx +. dy *. dy) in
      if dist > 4.0 then begin
        let nx = -. dy /. dist and ny = dx /. dist in
        control_pt <- Some (mx +. nx *. dist *. 0.3, my +. ny *. dist *. 0.3)
      end;
      ctx.request_update ()
    | None -> ()

  method on_release (ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(alt : bool) =
    ignore (shift, alt);
    match drag_start with
    | None -> ()
    | Some (sx, sy) ->
      drag_start <- None;
      drag_end <- None;
      let w = abs_float (x -. sx) and h = abs_float (y -. sy) in
      if w > 4.0 || h > 4.0 then begin
        let d = match control_pt with
          | Some (cx, cy) ->
            [Element.MoveTo (sx, sy); Element.CurveTo (cx, cy, cx, cy, x, y)]
          | None ->
            [Element.MoveTo (sx, sy); Element.LineTo (x, y)]
        in
        let elem = Element.make_text_path
          ~fill:(Some Element.{ fill_color = { r = 0.0; g = 0.0; b = 0.0; a = 1.0 } })
          d "Lorem Ipsum" in
        ctx.controller#add_element elem
      end;
      control_pt <- None

  method on_double_click (_ctx : Canvas_tool.tool_context) (_x : float) (_y : float) = ()
  method on_key (_ctx : Canvas_tool.tool_context) (_key : int) = false
  method activate (_ctx : Canvas_tool.tool_context) = ()
  method deactivate (_ctx : Canvas_tool.tool_context) = ()

  method draw_overlay (_ctx : Canvas_tool.tool_context) cr =
    match drag_start, drag_end with
    | Some (sx, sy), Some (ex, ey) ->
      Cairo.set_source_rgba cr 0.4 0.4 0.4 1.0;
      Cairo.set_line_width cr 1.0;
      Cairo.set_dash cr [| 4.0; 4.0 |];
      Cairo.move_to cr sx sy;
      (match control_pt with
       | Some (cx, cy) ->
         Cairo.curve_to cr cx cy cx cy ex ey
       | None ->
         Cairo.line_to cr ex ey);
      Cairo.stroke cr;
      Cairo.set_dash cr [||]
    | _ -> ()
end
